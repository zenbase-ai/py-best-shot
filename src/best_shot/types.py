from dataclasses import dataclass
from functools import cached_property
from hashlib import sha256
from typing import TypeVar

import ujson


IO = TypeVar("IO", bound=dict | str)
Datum = TypeVar("Datum", bound=tuple[IO, IO] | tuple[IO, IO, str])


def is_io_value(value) -> bool:
    return isinstance(value, (dict, str))


def data_key(data: dict) -> str:
    return ujson.dumps(data, sort_keys=True)


def data_hash(data: dict) -> str:
    return sha256(data_key(data).encode()).hexdigest()


@dataclass
class Shot:
    inputs: IO
    outputs: IO
    id: str

    def __init__(self, inputs: IO, outputs: IO, id: str = ""):
        super().__init__()
        self.inputs = inputs
        self.outputs = outputs
        self.id = id or data_hash(inputs)

    @cached_property
    def key(self) -> str:
        return data_key(self.inputs)


ShotWithSimilarity = TypeVar("ShotWithSimilarity", bound=tuple[Shot, float])
