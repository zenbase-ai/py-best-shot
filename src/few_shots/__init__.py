from .async_client import AsyncFewShots
from .client import FewShots
from .types import Shot, Embedding, ShotWithSimilarity, IO, Datum
from .utils.format import shots_to_messages

__all__ = [
    "AsyncFewShots",
    "FewShots",
    "Datum",
    "Embedding",
    "IO",
    "Shot",
    "shots_to_messages",
    "ShotWithSimilarity",
]
