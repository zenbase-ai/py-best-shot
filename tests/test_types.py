from few_shots.types import Shot, dump_io_value, id_io_value, is_io_value


def test_is_io_value():
    assert is_io_value("test") is True
    assert is_io_value({"key": "value"}) is True
    assert is_io_value(123) is False
    assert is_io_value(["test"]) is False


def test_data_key():
    data = {"b": 2, "a": 1}
    assert dump_io_value(data) == '{"a":1,"b":2}'


def test_data_hash(snapshot):
    data = {"test": "value"}
    assert id_io_value(data) == snapshot


def test_init_with_id():
    shot = Shot({"input": "test"}, {"output": "result"}, "test_id")
    assert shot.inputs == {"input": "test"}
    assert shot.outputs == {"output": "result"}
    assert shot.id == "test_id"


def test_init_without_id():
    inputs = {"input": "test"}
    shot = Shot(inputs, {"output": "result"})
    assert shot.id == id_io_value(inputs)


def test_key_property():
    inputs = {"input": "test"}
    shot = Shot(inputs, {"output": "result"})
    assert shot.key == dump_io_value(inputs)
