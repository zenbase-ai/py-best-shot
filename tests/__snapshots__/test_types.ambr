# serializer version: 1
# name: test_data_hash
  '928c36c9-b984-5597-9fc6-b20eafa55115'
# ---
